katrain: init at v1.17.0

iofq opened the pull request:
Adding [katrain](https://github.com/sanderland/katrain) as it's becoming difficult to build on a modern Linux system such as ubuntu and nix could help :)

To do so, I had to add 2 new python modules, kivymd and ffpyplayer. ffpyplayer was difficult because the project needs ffmpeg C bindings; after troubleshooting for a few hours, I still wasn't able to get it to compile against any version of ffmpeg, and fell back to the wheel from pypi (which appears to be built with ffmpeg 6).

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [ ] ** Only tested x86_64, do not have a darwin machine **
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

iofq committed:
python312Packages.ffpyplayer: init at 4.5.3

iofq committed:
python312Packages.kivymd: init at 1.1.1

iofq committed:
katrain: init at 1.17

iofq committed:
maintainers: add iofq to maintainers-list

pilz0 commented on pkgs/development/python-modules/ffpyplayer/default.nix:
```diff
@@ -0,0 +1,30 @@
+  version = "v4.5.2";
+  format = "wheel";
+
+  src = fetchurl {
```
I think it would be better to use the packages directly from their github: https://github.com/matham/ffpyplayer/releases/tag/v4.5.2

iofq replied:
makes sense, easier to bump versions in the future. updated + linted commits

pilz0 reviewed (commented):
Looks okay to me
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 10 packages built:</summary>
  <ul>
    <li>katrain</li>
    <li>katrain.dist</li>
    <li>python312Packages.ffpyplayer</li>
    <li>python312Packages.ffpyplayer.dist</li>
    <li>python312Packages.kivymd</li>
    <li>python312Packages.kivymd.dist</li>
    <li>python313Packages.ffpyplayer</li>
    <li>python313Packages.ffpyplayer.dist</li>
    <li>python313Packages.kivymd</li>
    <li>python313Packages.kivymd.dist</li>
  </ul>
</details>

pilz0 commented on pkgs/development/python-modules/ffpyplayer/default.nix:
```diff
@@ -0,0 +1,32 @@
+      else
+        "sha256-+6aibWImMvNe6GNP/JSos0QQJq8NZ7X4v+MwpC4BPFw=";
+  };
+
```
im pretty sure it would fail to build on linux/aarch64, so it would be good to build in support for arm64, with something like this.  I will test later test if it builds on aarch64.
```
    url =
      if stdenv.isDarwin then
        "https://github.com/matham/ffpyplayer/releases/download/v${version}/ffpyplayer-${version}-cp313-cp313-macosx_10_13_universal2.whl"
      if stdenv.isAarch64 then
        "/link/to/arch64.whl"
      else
        "https://github.com/matham/ffpyplayer/releases/download/v${version}/ffpyplayer-${version}-cp313-cp313-manylinux_2_17_x86_64.manylinux2014_x86_64.whl";

    hash =
      if stdenv.isDarwin then
        "sha256-mPvggPi8KbOCuLqHxxHPcvqSt3BGWVgsLlTIFU8mlsU="
      if stdenv.isAarch64 then
        "sha256-Foooooooooooooooooooooooooooooooooooooooo="

      else
        "sha256-+6aibWImMvNe6GNP/JSos0QQJq8NZ7X4v+MwpC4BPFw=";
  };
```

iofq replied:
unfortunately the project doesn't offer linux arm64 wheels, and they have a [custom](https://github.com/matham/ffpyplayer/tree/95562a514e63287b3eef3258a88ccf3134f1d369/.ci) CI setup to build deps that I haven't had luck replicating in nix and would rather not chase. I could open an issue over there to start supporting arm64 wheels and circle back to nixpkgs when that's complete, if that's an acceptable workaround?

iofq replied:
I'm working on a PR to add arm64 wheels over at ffpyplayer and will circle back here when those are available.

iofq replied:
@pilz0 I've added ffpyplayer arm64 wheels upstream and referenced them in this PR, would you like to take another look?

pilz0 replied:
Hey, thanks for adding arm wheel builds upstream, im going to to take a look at the PR later today. Im also pretty new to nixpkgs so i cant find all issues.

pilz0 commented:
Hey, ive looked over the PR again, and at least for me KivyMD doesnt work (amd64). ffpyplayer seems to work on both arm and amd64. Otherwise i haved noticed any issues :)

pilz0 commented:
> > Hey, ive looked over the PR again, and at least for me KivyMD doesnt work (amd64). ffpyplayer seems to work on both arm and amd64. Otherwise i haved noticed any issues :)
> 
> Hey, can you describe what exactly is not working for kivymd? For me it builds and katrain can use it fine.

Im getting this error when im running katrain via nixpkgs-review
``
[ERROR  ] [Clock       ] Unable to import kivy._clock. Have you perhaps forgotten to compile kivy? Kivy contains Cython code which needs to be compiled. A missing kivy._clock often indicates the Cython code has not been compiled. Please follow the installation instructions and make sure to compile Kivy
 Traceback (most recent call last):
   File "/nix/store/f61x1pl8rhhwv10ddj5c2jcn864k1dik-katrain-v1.16/bin/.katrain-wrapped", line 6, in <module>
     from katrain.__main__ import run_app
   File "/nix/store/f61x1pl8rhhwv10ddj5c2jcn864k1dik-katrain-v1.16/lib/python3.12/site-packages/katrain/__main__.py", line 49, in <module>
     from kivy.base import ExceptionHandler, ExceptionManager
   File "/nix/store/g5na9p8x5r0qfyz98rq1n3rbgbfw5xxf-python3.13-kivy-2.3.1/lib/python3.13/site-packages/kivy/base.py", line 28, in <module>
     from kivy.clock import Clock
   File "/nix/store/g5na9p8x5r0qfyz98rq1n3rbgbfw5xxf-python3.13-kivy-2.3.1/lib/python3.13/site-packages/kivy/clock.py", line 466, in <module>
     from kivy._clock import CyClockBase, ClockEvent, FreeClockEvent, \
 ModuleNotFoundError: No module named 'kivy._clock
``

pilz0 commented:
Python gives me this error:
```
Python 3.12.10 (main, Apr  8 2025, 11:35:47) [GCC 14.2.1 20250322] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import kivymd
[INFO   ] [Logger      ] Record log in /home/marie/.kivy/logs/kivy_25-06-13_1.txt
[INFO   ] [Kivy        ] v2.3.1
[INFO   ] [Kivy        ] Installed at "/nix/store/g5na9p8x5r0qfyz98rq1n3rbgbfw5xxf-python3.13-kivy-2.3.1/lib/python3.13/site-packages/kivy/__init__.py"
[INFO   ] [Python      ] v3.12.10 (main, Apr  8 2025, 11:35:47) [GCC 14.2.1 20250322]
[INFO   ] [Python      ] Interpreter at "/nix/store/8w718rm43x7z73xhw9d6vh8s4snrq67h-python3-3.12.10/bin/python"
[INFO   ] [Logger      ] Purge log fired. Processing...
[INFO   ] [Logger      ] Purge finished!
[INFO   ] [KivyMD      ] 1.1.1, git-Unknown, 2022-10-13 (installed at "/nix/store/5prvh59y4rp41bsvqyv3h0kdrpi2i5by-python3.13-kivymd-1.1.1/lib/python3.13/site-packages/kivymd/__init__.py")
[INFO   ] [Factory     ] 195 symbols loaded
 Traceback (most recent call last):
   File "<stdin>", line 1, in <module>
   File "/nix/store/5prvh59y4rp41bsvqyv3h0kdrpi2i5by-python3.13-kivymd-1.1.1/lib/python3.13/site-packages/kivymd/__init__.py", line 66, in <module>
     import kivymd.font_definitions  # NOQA
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   File "/nix/store/5prvh59y4rp41bsvqyv3h0kdrpi2i5by-python3.13-kivymd-1.1.1/lib/python3.13/site-packages/kivymd/font_definitions.py", line 10, in <module>
     from kivy.core.text import LabelBase
   File "/nix/store/g5na9p8x5r0qfyz98rq1n3rbgbfw5xxf-python3.13-kivy-2.3.1/lib/python3.13/site-packages/kivy/core/text/__init__.py", line 85, in <module>
     from kivy.graphics.texture import Texture
   File "/nix/store/g5na9p8x5r0qfyz98rq1n3rbgbfw5xxf-python3.13-kivy-2.3.1/lib/python3.13/site-packages/kivy/graphics/__init__.py", line 89, in <module>
     from kivy.graphics.instructions import Callback, Canvas, CanvasBase, \
 ModuleNotFoundError: No module named 'kivy.graphics.instructions'
```

iofq committed:
python312Packages.asyncgui: init at 0.9.3

iofq committed:
python312Packages.asynckivy: init at 0.9.3

iofq committed:
nixos/katrain: init module
Adds a module for pkgs.katrain to govern katago patching behavior and additional dependencies

EinEtw4s commented on pkgs/by-name/ka/katrain/package.nix:
```diff
@@ -0,0 +1,74 @@
+    owner = "sanderland";
+    repo = "katrain";
+    tag = version;
+    hash = "sha256-v0VGjxayXLbBSXs/jkYeXgbyDg+JFLgxESDuh/qbg1U=";
```
I'm getting a different hash here

```suggestion
    hash = "sha256-2DHamhxsOPKldRVxKHUoGKZn1Ao9l4D+TB4NkroixbA=";
```

iofq replied:
thanks, I had this cached somehow from when i bumped 1.16 -> 1.17

EinEtw4s commented:
builds for me now, too, without error

iofq commented:
Hello @EinEtw4s @pilz0,

Can you describe how you're reaching the import error in https://github.com/NixOS/nixpkgs/pull/408590#issuecomment-2969370882?

I'm using the following flake to test importing each python package without issue:

```nix
{
  inputs = {
    nixpkgs.url = "github:iofq/nixpkgs/init_katrain";
  };

  outputs = {
    self,
    nixpkgs,
  }: {
    devShells.x86_64-linux.default = let
      pkgs = import nixpkgs {
        system = "x86_64-linux";
      };
    in
      pkgs.mkShell {
        packages = [
          pkgs.katrain
          (
            pkgs.python3.withPackages
            (ps:
              with ps; [
                kivymd
                asyncgui
                asynckivy
                ffpyplayer
              ])
          )
        ];
      };
  };
}
```

```bash
iofq@nix ~/d/nixreview> nix develop --command python -c "import asyncgui,asynckivy,kivymd,ffpyplayer"
[INFO   ] [Kivy        ] v2.3.1
[INFO   ] [Kivy        ] Installed at "/nix/store/a0y8hih1b2c2n0jpy7agk3kxrbfdikgk-python3.13-kivy-2.3.1/lib/python3.13/site-packages/
kivy/__init__.py"
[INFO   ] [Python      ] v3.13.5 (main, Jun 11 2025, 15:36:57) [GCC 14.3.0]
[INFO   ] [Python      ] Interpreter at "/nix/store/djck7mx6jad1w0yy6zings96dyxanls6-python3-3.13.5/bin/python"
[INFO   ] [Logger      ] Purge log fired. Processing...
[INFO   ] [Logger      ] Purge finished!
[INFO   ] [Image       ] Providers: img_tex, img_dds, img_sdl2, img_pil (img_ffpyplayer ignored)
[INFO   ] [Factory     ] 195 symbols loaded
[INFO   ] [Window      ] Provider: sdl2
[INFO   ] [GL          ] Using the "OpenGL" graphics system
[INFO   ] [GL          ] Backend used <sdl2>
[INFO   ] [GL          ] OpenGL version <b'4.6 (Compatibility Profile) Mesa 25.1.1'>
[INFO   ] [GL          ] OpenGL vendor <b'AMD'>
[INFO   ] [GL          ] OpenGL renderer <b'AMD Radeon Graphics (radeonsi, renoir, ACO, DRM 3.63, 6.15.0)'>
[INFO   ] [GL          ] OpenGL parsed version: 4, 6
[INFO   ] [GL          ] Shading version <b'4.60'>
[INFO   ] [GL          ] Texture max size <16384>
[INFO   ] [GL          ] Texture max units <32>
[INFO   ] [Window      ] auto add sdl2 input provider
[INFO   ] [Window      ] virtual keyboard not allowed, single mode, not docked
[INFO   ] [KivyMD      ] v0.104.1
```

iofq commented:
Hello @eetwas @pilz0 

I've rebased and updated versions on this, and it's still passing all my testing, including the flake test in https://github.com/NixOS/nixpkgs/pull/408590#issuecomment-3105503519

Would you be willing to review this again so we can get it to the finish line?

Much appreciated!
